patches/server/1048-add-ItemFrameMeta-API.patch

```diff
@@ -0,0 +1,323 @@
++        Material.ITEM_FRAME,
++        Material.GLOW_ITEM_FRAME
++    );
++
++    static final ItemMetaKeyType<CustomData> ENTITY_TAG = new ItemMetaKeyType<>(DataComponents.ENTITY_DATA, "EntityTag", "entity-tag");
++    CompoundTag entityTag;
++
++    static final ItemMetaKey ENTITY_ID = new ItemMetaKey("id", "entity-id");
++    static final ItemMetaKey FIXED = new ItemMetaKey("Fixed", "fixed");
++    static final ItemMetaKey INVISIBLE = new ItemMetaKey("Invisible", "invisible");
++
++    CraftMetaItemFrame(CraftMetaItem meta) {
++        super(meta);
++
++        if (meta instanceof CraftMetaItemFrame itemFrame) {
++            this.entityTag = itemFrame.entityTag;
++        }
++    }
++
++    CraftMetaItemFrame(DataComponentPatch tag, final Set<DataComponentType<?>> extraHandledDcts) {
++        super(tag, extraHandledDcts);
++
++        getOrEmpty(tag, CraftMetaItemFrame.ENTITY_TAG).ifPresent((nbt) -> this.entityTag = nbt.copyTag());
++    }
++
++    CraftMetaItemFrame(Map<String, Object> map) {
++        super(map);
++
++        String entityTag = SerializableMeta.getString(map, ENTITY_TAG.BUKKIT, true);
++        if (entityTag != null) {
++            ByteArrayInputStream buf = new ByteArrayInputStream(Base64.getDecoder().decode(entityTag));
++            try {
++                this.entityTag = NbtIo.readCompressed(buf, NbtAccounter.unlimitedHeap());
++            } catch (IOException ex) {
++                Logger.getLogger(CraftMetaItem.class.getName()).log(Level.SEVERE, null, ex);
++            }
++            return;
++        }
++        SerializableMeta.getObjectOptionally(Boolean.class, map, FIXED.BUKKIT, true).ifPresent((value) -> {
++            populateTagIfNull();
++            this.entityTag.putBoolean(FIXED.NBT, value);
++        });
++        SerializableMeta.getObjectOptionally(Boolean.class, map, INVISIBLE.BUKKIT, true).ifPresent((value) -> {
++            populateTagIfNull();
++            this.entityTag.putBoolean(INVISIBLE.NBT, value);
++        });
++    }
++
++    @Override
++    void deserializeInternal(CompoundTag tag, Object context) {
++        super.deserializeInternal(tag, context);
++
++        if (tag.contains(CraftMetaItemFrame.ENTITY_TAG.NBT)) {
++            this.entityTag = tag.getCompound(CraftMetaItemFrame.ENTITY_TAG.NBT);
++
++            // fix up legacy item frame metas that did not include entity ID.
++            if (!this.entityTag.contains(ENTITY_ID.NBT)) {
++                this.entityTag.putString(ENTITY_ID.NBT, "minecraft:item_frame");
++            }
++        }
++    }
++
++    @Override
++    void serializeInternal(Map<String, Tag> internalTags) {
++        // correctly serialised as entity tag.
++    }
++
++    @Override
++    protected void applyToItem(CraftMetaItem.Applicator tag) {
++        super.applyToItem(tag);
++
++        if (this.entityTag != null) {
++            tag.put(CraftMetaItemFrame.ENTITY_TAG, CustomData.of(this.entityTag));
++        }
```

Comment:
We probably need to do a sanity check here to make sure the custom data compound, if its not empty, that it has the "id" key/value in it. If it doesn't and its not empty, it'll throw errors all over the place as the codec requires it.